modes/vi-mode/core.lisp

```diff
@@ -55,27 +59,50 @@
   (setf (element-name *modeline-element*) name))
 
 
-(defstruct vi-state
-  name
-  keymap
-  post-command-hook
-  enable-hook
-  disable-hook
-  cursor-color)
+(defclass vi-state ()
+  ((tag
+  :initarg :tag
+  :reader tag)
+  (message
+  :initarg :message
+  :reader msg)
+  (cursor-type 
+  :initarg :cursor-type
+  :reader cursor-type)
+  (keymap
+  :initarg :keymap
+  :reader state-keymap)
+  (cursor-color
+  :initarg :cursor-color
+  :accessor cursor-color)))
 
 (defvar *current-state* nil)
 
-(defmacro define-vi-state (name (&key keymap post-command-hook cursor-color) &body spec)
-  (let ((enable-form (rest (assoc :enable spec)))
-        (disable-form (rest (assoc :disable spec))))
-    `(setf (get ',name 'state)
-           (make-vi-state :name ',name
+;;; vi-state methods
+(defmacro define-vi-state (name (&key tag message cursor-type keymap cursor-color) &body spec)
+    `(progn
+      (cl-package-locks:without-package-locks
+       (defclass ,name (vi-state) ()))
```

Comment:
Since define a class with the same name when defining command and mode, is there any problem that conflicts with it?

https://github.com/lem-project/lem/blob/fa4356714e49c61e6fb7094bd7f705f994df3233/src/defcommand.lisp#L109

Reply:
I'll need to investigate this. It may be an issue if the names are the same, assuming the symbols exported into the global package space.

Reply:
I tested this in using the `define-command` to define a conflicting `insert` `lem` `command` within the main package. Which in theory would conflict with `insert` `vi-state` in the `vi-mode` package; however, it doesn't produce conflicts. In addition, the `vi-states` defined are not exported by the package. However, feel free to test, and provide feedback or findings. https://github.com/lem-project/lem/blob/fa4356714e49c61e6fb7094bd7f705f994df3233/modes/vi-mode/vi-mode.lisp#L6-L10

Reply:
Perhaps there is a conflict with the lem:attribute symbol?

Reply:
I fixed the name collision. Thank you for catching this.